File: ts/src/values/valuePropertyAccess.ts
import { IArrayValue, isIArrayValue, isIObjectValue, IValue } from "./ivalues";
import NumberValue from "./numberValue";
import StringValue from "./stringValue";
import VariableValue from "./variableValue";

export function getProperty(current: IValue, properties: IArrayValue): IValue | undefined
{
    const propValues = properties.arrayValues();
    for (let i = 0; i < propValues.length; i++)
    {
        const index = parseIndex(propValues[i]);
        if (index !== undefined && isIArrayValue(current))
        {
            const test = current.tryGetIndex(index);
            if (test === undefined)
            {
                return undefined;
            }
            current = test;
        }
        else if (isIObjectValue(current))
        {
            const test = current.tryGetKey(propValues[i].toString());
            if (test === undefined)
            {
                return undefined;
            }
            current = test;
        }

        return undefined;
    }

    return current;
}

export function parseIndex(input: IValue): number | undefined
{
    if (input instanceof NumberValue)
    {
        return input.value;
    }
    else if (input instanceof StringValue || input instanceof VariableValue)
    {
        const index = parseInt(input.value);
        return isFinite(index) ? index : undefined;
    }

    return undefined;
}

File: unity/lysithea-vm-unity/Assets/Scripts/LysitheaVM/Values/ValuePropertyAccess.cs
using System.Diagnostics.CodeAnalysis;

#nullable enable

namespace LysitheaVM
{
    public static class ValuePropertyAccess
    {
        public static bool TryGetProperty(IValue current, ArrayValue properties, [NotNullWhen(true)] out IValue? result)
        {
            result = NullValue.Value;
            for (var i = 0; i < properties.Length; i++)
            {
                if (TryParseIndex(properties[i], out var index) && current is IArrayValue currentArray)
                {
                    if (!currentArray.TryGetIndex(index, out var test))
                    {
                        return false;
                    }
                    current = test;
                }
                else if (current is IObjectValue currentObject)
                {
                    if (!currentObject.TryGetKey(properties[i].ToString(), out var test))
                    {
                        result = NullValue.Value;
                        return false;
                    }
                    current = test;
                }
                else
                {
                    return false;
                }
            }

            result = current;
            return true;
        }

        public static bool TryParseIndex(IValue input, out int result)
        {
            if (input is NumberValue numberValue)
            {
                result = numberValue.IntValue;
                return result >= 0;
            }
            else if (input is StringValue stringValue && int.TryParse(stringValue.Value, out result))
            {
                return true;
            }
            else if (input is VariableValue variableValue && int.TryParse(variableValue.Value, out result))
            {
                return true;
            }

            result = -1;
            return false;
        }
    }
}